main/src/library/Map.flix

```diff
@@ -1,7 +1,335 @@
+
+    ///
+    /// Updates `m` with `k -> v1` if `k -> v` is in `m` and `f(v) = Some(v1).
+    /// Otherwise, returns `m`.
+    ///
+    def update[k, v](f: v -> Option[v], k: k, m: Map[k, v]): Map[k, v] = updateWithKey((k1, v1) -> f(v1), k, m)
+
+    ///
+    /// Updates `m` with `k -> v1` if `k -> v` is in `m` and `f(k, v) = Some(v1).
+    /// Otherwise, returns `m`.
+    ///
+    def updateWithKey[k, v](f: (k, v) -> Option[v], k: k, m: Map[k, v]): Map[k, v] =
+        let Map(xs) = m;
+            Map(List.map(match (key, v) -> if (key == k) (k, Option.getWithDefault(f(k, v), v)) else (key, v), xs))
+
+    //
+    // ## Delete
+    //
+
+    ///
+    ///
+    ///
+    def delete[k, v](k: k, m: Map[k, v]): Map[k, v] =
+        let Map(xs) = m;
+            Map(List.filter(match (key, v) -> key != k, xs))
+
+    //
+    // ## Map Predicates
+    //
+
+    ///
+    /// Returns `true` if and only if all mappings in `m1` occur in `m2`.
+    ///
+    def isSubmapOf[k, v](m1: Map[k, v], m2: Map[k, v]): Bool =
+        let Map(xs) = m1;
+        let Map(ys) = m2;
+            List.forall(x -> List.memberOf(x, ys), xs)
+
+    ///
+    /// Returns `true` if and only if all mappings in `m1` occur in `m2` and `m1 != m2`.
+    ///
+    def isProperSubmapOf[k, v](m1: Map[k, v], m2: Map[k, v]): Bool = isSubmapOf(m1, m2) && size(m1) != size(m2)
+
+    //
+    // ## Map Transformation
+    //
+
+    ///
+    /// Returns a map of all mappings `k -> v` in `m` where `v` satisfies the predicate `f`.
+    ///
+    def filter[k, v](f: v -> Bool, m: Map[k, v]): Map[k, v] = filterWithKey((k, v) -> f(v), m)
+
+    ///
+    /// Returns a map of all mappings `k -> v` in `m` where `(k, v)` satisfies the predicate `f`.
+    ///
+    def filterWithKey[k, v](f: (k, v) -> Bool, m: Map[k, v]): Map[k, v] =
+        let Map(xs) = m;
+            Map(List.filter(x -> f(fst(x), snd(x)), xs))
+
+    ///
+    /// Returns a map with mappings `k -> f(v)` for every `k -> v` in `m`.
+    ///
+    def map[k, v1, v2](f: v1 -> v2, m: Map[k, v1]): Map[k, v2] = mapWithKey((k, v) -> f(v), m)
+
+    ///
+    /// Returns a map with mappings `k -> f(k, v)` for every `k -> v` in `m`.
+    ///
+    def mapWithKey[k, v1, v2](f: (k, v1) -> v2, m: Map[k, v1]): Map[k, v2] =
+        let Map(xs) = m;
+            Map(List.map(match (k, v) -> (k, f(k, v)), xs))
+
+    //
+    // ## Fold
+    //
+
+    ///
+    /// Alias for `foldLeft`.
+    ///
+    def fold[k, v, b](f: (b, v) -> b, s: b, m: Map[k, v]): b = foldLeft(f, s, m)
+
+    ///
+    /// Alias for `foldLeftWithKey`.
+    ///
+    def foldWithKey[k, v, b](f: (k, b, v) -> b, s: b, m: Map[k, v]): b = foldLeftWithKey(f, s, m)
+
+    ///
+    /// Applies `f` to a start value `s` and all values in `m` going from left to right.
+    /// That is, the result is of the form: `f(...f(f(s, v1), v2)..., vn)`.
+    ///
+    def foldLeft[k, v, b](f: (b, v) -> b, s: b, m: Map[k, v]): b = foldLeftWithKey((k, b, v) -> f(b, v), s, m)
```

Comment:
I'm guessing that in both `foldLeft` and `foldRight` the function `f` should be commutative (`f(a,b)=f(b,a)`) since the order of the key-value pairs in a map is not guaranteed.  If this is the case, what's the point of having both `foldLeft` and `foldRight` when order is not guaranteed?